### PR TITLE
gh-98086: Now ``patch.dict`` can decorate async functions

### PR DESCRIPTION
**Pull request by sobolevn**
I will check other `patch.*` types for similar problems in the next issues / PRs.

Issue: gh-98086

### CODE REVIEW

**Review by seifertm — APPROVED** [review]
Thanks for looking into this! The changes fix the reported issue: coroutines now stay coroutines when applying the `patch.dict` decorator.

I checked the behavior of all `patch.*` functions and they seem to be consistent as of this PR. Patching preserves coroutines, even inside of `TestCase` classes when used as a class decorator.

Note that async generator functions are _not_ preserved. They become functions when decorated with `patch.*`. However, this is the same for all variants of patch and not part of the reported issue.

Here's the code I used to verify the behavior:
```
import inspect
from unittest import TestCase
from unittest.mock import DEFAULT, Mock, patch


patchers = (
    patch("binascii.b64encode"),
    patch.dict("sys.modules", test_module=Mock()),
    patch.object(Mock, "my_func"),
    patch.multiple(Mock, my_attr=DEFAULT, my_other=DEFAULT),
)

async def coro():
    ...

patched_coros = (patcher(coro) for patcher in patchers)

assert inspect.iscoroutinefunction(coro)
assert all(map(inspect.iscoroutinefunction, patched_coros))


# Coroutines in test classes
for patcher in patchers:
    @patcher
    class TestClass(TestCase):
        async def test_coro(self):
            ...
    assert inspect.iscoroutinefunction(TestClass.test_coro)



# The behavior across patch functions is consistent with async generator functions
async def async_gen_func():
    yield

patched_async_gen_funcs = (patcher(async_gen_func) for patcher in patchers)
assert inspect.isasyncgenfunction(async_gen_func)
assert not all(map(inspect.isasyncgenfunction, patched_async_gen_funcs))
```

**Comment by seifertm** [comment]
@sobolevn Will these changes be considered for backports to earlier Python versions?

I'm asking, because I'm trying to assess if pytest-asyncio needs to provide a workaround.

**Comment by sobolevn** [comment]
This looks like a bugfix to me. So, according to the regular policy: yes, they will be.
But, there might be exceptions. Let's wait for the code owner :)

**Review by changguoMA — APPROVED** [review]

**Comment by sobolevn** [comment]
@cjw296 friendly ping :)

**Comment by cjw296** [comment]
I'm afraid I don't have time to review this right now. @tirkarthi / @mariocj89 ? 

@seifertm - even if merged, this will only be backported to 3.11 and 3.10.

**Review by tirkarthi — APPROVED** [review]
LGTM, Thanks @sobolevn . This looks similar to the logic already done for patch and now adopted to `patch.dict` too. I will wait for review from @mariocj89 and will try to merge it within end of this week along with backport.

https://github.com/python/cpython/blob/a751bf565c731c87f2e02678ee080f04c86d7eee/Lib/unittest/mock.py#L1324-L1326

**Review by mariocj89 — APPROVED** [review]
Makes sense. We missed this when we updated the `_patch` function indeed.

**Comment by sobolevn** [comment]
Thanks everyone, I've scheduled 3.10 and 3.11 backports

File: Misc/NEWS.d/next/Library/2022-10-08-19-39-27.gh-issue-98086.y---WC.rst
Comment (merwok) [inline Misc/NEWS.d/next/Library/2022-10-08-19-39-27.gh-issue-98086.y---WC.rst:1]: For future PRs, please make sure that the news fragment makes sense when it’s read in the compiled changelog.
Here `patch` is mentioned but there is no such builtin, it would be better to name `unittest.mock.patch.dict` (like the issue helpfully did).  Thanks!

**Pull request merged**
